src/node.cc

```diff
@@ -3817,22 +3799,73 @@ int Start(int argc, char** argv) {
           more = true;
       }
     } while (more == true);
-    code = EmitExit(env);
+
+    int exit_code = EmitExit(env);
+    if (instance_data->is_main())
+      instance_data->set_exit_code(exit_code);
     RunAtExit(env);
 
     env->Dispose();
     env = nullptr;
   }
 
-  CHECK_NE(node_isolate, nullptr);
-  node_isolate->Dispose();
-  node_isolate = nullptr;
+  CHECK_NE(isolate, nullptr);
+  isolate->Dispose();
+  isolate = nullptr;
+  if (instance_data->is_main())
+    node_isolate = nullptr;
+
+  // Main node instance uses uv_default_loop.
+  if (instance_data->is_worker())
+    uv_loop_delete(instance_data->event_loop());
```

Comment:
Looking at this again, I think it should be left to the caller to clean up the uv_loop_t.  Calling uv_loop_delete() is not safe because we don't know if the loop was created with uv_loop_new() or uv_loop_init().